adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/LoraProtocolAdapter.java

```diff
@@ -0,0 +1,636 @@
+                getConfig().getRealm(), tracer));
+
+        router.route().handler(authHandler);
+    }
+
+    void handleProviderRoute(final RoutingContext ctx, final LoraProvider provider) {
+        LOG.debug("Handling route for provider with path: [{}]", provider.pathPrefix());
+        final Span currentSpan = getCurrentSpan(ctx);
+        ctx.put(LoraConstants.APP_PROPERTY_ORIG_LORA_PROVIDER, provider.getProviderName());
+        currentSpan.setTag(TAG_LORA_PROVIDER, provider.getProviderName());
+
+        if (ctx.user() instanceof Device) {
+            final Device gatewayDevice = (Device) ctx.user();
+            final JsonObject loraMessage = ctx.getBodyAsJson();
+            currentSpan.setTag(MessageHelper.APP_PROPERTY_TENANT_ID, gatewayDevice.getTenantId());
+
+            LoraMessageType type = LoraMessageType.UNKNOWN;
+            try {
+                type = provider.extractMessageType(loraMessage);
+                final String deviceId = provider.extractDeviceId(loraMessage);
+                currentSpan.setTag(TAG_LORA_DEVICE_ID, deviceId);
+                currentSpan.setTag(MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
+
+                if (LoraMessageType.UPLINK.equals(type)) {
+                    final String payloadBase64 = provider.extractPayloadEncodedInBase64(loraMessage);
+                    doUpload(ctx, gatewayDevice, deviceId, payloadBase64);
+                } else {
+                    LOG.debug("Received message '{}' of type [{}] for device [{}], will discard message.", loraMessage,
+                            type, deviceId);
+                    currentSpan.log(
+                            "Received message of type '" + type + "' for device '" + deviceId + "' will be discarded.");
+                    // Throw away the message but return 202 to not cause errors on the LoRa provider side
+                    handle202(ctx);
+                }
+            } catch (final ClassCastException | LoraProviderMalformedPayloadException e) {
+                LOG.debug("Got invalid payload '{}' which leads to exception: {}", loraMessage, e);
+                TracingHelper.logError(currentSpan,
+                        "Received message of type '" + type + "' has invalid payload; error: " + e);
+                handle400(ctx, INVALID_PAYLOAD);
+            }
+        } else {
+            LOG.debug("Supplied credentials are not an instance of the user. Returning 401");
+            TracingHelper.logError(currentSpan, "Supplied credentials are not an instance of the user");
+            handle401(ctx);
+        }
+    }
+
+    private Span getCurrentSpan(final RoutingContext ctx) {
+        return (ctx.get(TracingHandler.CURRENT_SPAN) instanceof Span) ? ctx.get(TracingHandler.CURRENT_SPAN)
+                : NoopSpan.INSTANCE;
+    }
+
+    void handleOptionsRoute(final RoutingContext ctx) {
+        LOG.debug("Handling options method");
+
+        if (ctx.user() instanceof Device) {
```

Comment:
IMHO this check does not add much value because you never use the `Device` data. The request will already have been rejected with a 401 if the provider didn't provide valid credentials, won't it?